ports/nrf/modules/machine/uart.c

```diff
@@ -70,27 +78,39 @@ STATIC int uart_find(mp_obj_t id) {
     mp_raise_ValueError("UART doesn't exist");
 }
 
-void uart_irq_handler(mp_uint_t uart_id) {
-
+void m_rx_handler(nrfx_uart_event_t const *p_event, void *p_context) {
+    machine_hard_uart_obj_t *self = p_context;
+    if (p_event->type == NRFX_UART_EVT_RX_DONE) {
+        int chr = rx_buf[0];
+        nrfx_uart_rx(self->p_uart, &rx_buf[0], 1);
+        #if !MICROPY_PY_BLE_NUS && MICROPY_KBD_EXCEPTION
+        if (chr == mp_interrupt_char) {
+            mp_keyboard_interrupt();
+        } else
+        #endif
+        {
+            ringbuf_put((ringbuf_t*)&rx_ringbuf, chr);
+        }
+    }
+    else if (p_event->type == NRFX_UART_EVT_TX_DONE) {
+        tx_pending = false;
+    }
 }
 
 bool uart_rx_any(const machine_hard_uart_obj_t *uart_obj) {
-    // TODO: uart will block for now.
-    return true;
+    return rx_ringbuf.iput != rx_ringbuf.iget;
 }
 
 int uart_rx_char(const machine_hard_uart_obj_t * self) {
-    uint8_t ch;
-    nrfx_uart_rx(self->p_uart, &ch, 1);
-    return (int)ch;
+    return ringbuf_get((ringbuf_t*)&rx_ringbuf);
 }
 
 STATIC nrfx_err_t uart_tx_char(const machine_hard_uart_obj_t * self, int c) {
-    while (nrfx_uart_tx_in_progress(self->p_uart)) {
-        ;
+    while (tx_pending) {
     }
-
-    return nrfx_uart_tx(self->p_uart, (uint8_t *)&c, 1);
+    tx_buf[0] = c;
+    tx_pending = true;
```

Comment:
The `nrfx_uart_tx_in_progress()` call seems to do what I want here but it didn't work, it always blocked.  So that's why I added the `tx_pending` flag and busy wait on it (instead of calling `nrfx_uart_tx_in_progress()`).

Reply:
Yeah, this one was an interesting one. I replaced your code back to `nrfx_uart_tx_in_progress()` and reproduced the issue. However, compiling the code with `DEBUG=1` and` LTO=0` made it work. This hints on a `volatile` bug. So, i updated the `uart_control_block_t.tx_buffer_length` to become a volatile, then it started to work. I'll follow up on the nrfx issue, meanwhile that means your code here is correct as you have moved this logic to micropython and kept it as a volatile. :)

Reply:
Good old LTO...

Reply:
The fix i sent to nrfx regarding `nrf_uart_tx_in_progress()` is now in. I have created the PR #4802 upgrading the nrfx to include the fix. It should by this new fix be possible to run the function without getting blocked.

Reply:
Great, thanks.  I changed it back to use `nrfx_uart_tx_in_progress` and indeed it works.